### DISCUSSION TITLE
Inaccurate postition estimation with TWR + custom UWB packets

### DISCUSSION THREAD

**Discussion by shossains** [bitcraze/discussions#1292] [Q&A]
Hi all,

I am currently working on a project where I try to fly multiple drones simultaneously using the Loco Positioning System.
Additionally, I want to **send current location of my drones over UWB to each other** as a form of P2P communication (lets call these packets _POS packets_ from now on). I have stumbled upon the problem when I send these POS packets over UWB alternatingly with the original TWR packets, the position estimation is significantly worsened.

In my current implementation I alternate 50-50 with sending TWR packets and POS packets. In the [TWR algorithm](https://github.com/bitcraze/crazyflie-firmware/blob/master/src/deck/drivers/src/lpsTwrTag.c#L417) I added a boolean which toggles on and off after each iteration. Since the rate is basically halved, my initial assumption was that not enough TWR were sent for an accurate enough position estimation. However, after further testing of 10-1, 20-1 and even a 100-1 TWR-POS packet ratio it still gave similair results in inaccurate position estimation. 

My questions are:
What is causing the position estimation to be so inaccurate when also sending my custom POS packets?
Is there some way of increasing the frequency at which the TWR are being send?
Is there a better way in approaching this?

Any help is appreciated

**Comment by knmcguire** [comment]
I think it will be quite difficult to debug your issue if it is based on modified firmware.

 However I could propose for you to look at an alternative. We have an experimental feature of Loco pos that is called hybrid TDOA, which are TDOA + P2P messages. Perhaps take a look at that instead?

Here is the blogpost: https://www.bitcraze.io/2023/11/tdoa3-hybrid-mode/

**Comment by shossains** [comment]
Thank you for you response!

I did check out hybrid mode and took some inspiration from it to improve my implementation, thanks!
After some more debugging and testing I noticed that the position estimation from the LPS actually happens when I turn on more than one crazyflie. I have even flashed stock firmware on the crazyflies (cf2_defconfig, v2023.11 +105) and this behaviour still occurs. My setup consists of a cage with 8 anchors spaced more than 2 meters apart from each other.

The following happens:
1. Turning on one crazyflie: the position estimate is accurate.
2. Turning on another crazyflie: the position estimate is all over the place for both crazyflies.

I'm not sure if I should make a seperate thread for this. But is this a common problem? Since I am using default firmware and config on the crazyflies, the anchors can be only other cause for this. If I recall correctly there is very little configuration on the anchor possible that might cause this behaviour, thus I am a bit lost on how to find the origin of this problem and how to fix it.

**Reply by knmcguire** [reply]
Oke so if this is not related to the hybrid mode and happens with the stock firmware (although we do prefer you try out the latest release which is 2024.02), then this would be better as a seperate thread. We haven't seen this behavior before with more crazyflies. 

Btw, I also realize that hybrid tdoa experimental mode has been implented into the latest firmware already https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/functional-areas/loco-positioning-system/tdoa3_hybrid_mode/. But things still need to be configured in order to get it to work.
